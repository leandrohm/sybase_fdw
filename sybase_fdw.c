/*-------------------------------------------------------------------------
 *
 * sybase_fdw.c
 *		foreign-data wrapper for Sybase.
 *
 * IDENTIFICATION
 *		contrib/sybase_fdw/sybase_fdw.c
 *
 *-------------------------------------------------------------------------
 */
#include "postgres.h"

#include "access/reloptions.h"
#include "catalog/pg_foreign_table.h"
#include "commands/copy.h"
#include "commands/defrem.h"
#include "commands/explain.h"
#include "foreign/fdwapi.h"
#include "foreign/foreign.h"
#include "miscadmin.h"
#include "optimizer/cost.h"

PG_MODULE_MAGIC;

/*
 * SQL functions
 */
extern Datum sybase_fdw_handler(PG_FUNCTION_ARGS);
extern Datum sybase_fdw_validator(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(sybase_fdw_handler);
PG_FUNCTION_INFO_V1(sybase_fdw_validator);

/*
 * Foreign-data wrapper handler function: return a struct with pointers
 * to my callback routines.
 */
Datum
sybase_fdw_handler(PG_FUNCTION_ARGS)
{
	FdwRoutine *fdwroutine = makeNode(FdwRoutine);

	/* TODO: implement them all ASAP */
	fdwroutine->PlanForeignScan = NULL;
	fdwroutine->ExplainForeignScan = NULL;
	fdwroutine->BeginForeignScan = NULL;
	fdwroutine->IterateForeignScan = NULL;
	fdwroutine->ReScanForeignScan = NULL;
	fdwroutine->EndForeignScan = NULL;

	PG_RETURN_POINTER(fdwroutine);
}

/*
 * Validate the generic options given to a FOREIGN DATA WRAPPER, SERVER,
 * USER MAPPING or FOREIGN TABLE that uses sybase_fdw.
 *
 * Raise an ERROR if the option or its value is considered invalid.
 */
Datum
sybase_fdw_validator(PG_FUNCTION_ARGS)
{
	/* dummy function ATM */
	/* TODO: implement this ASAP */

	PG_RETURN_VOID();
}

/* vim: set tabstop=4 shiftwidth=4 noexpandtab: */
