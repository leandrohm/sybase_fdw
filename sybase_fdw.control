# sybase_fdw extension
comment = 'foreign-data wrapper for Sybase'
default_version = '1.0'
module_pathname = '$libdir/sybase_fdw'
relocatable = true
