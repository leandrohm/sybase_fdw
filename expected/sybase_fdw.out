--
-- Test Sybase Foreign Data Wrapper
--
-- Clean up in case a prior regression run failed
SET client_min_messages TO 'error';
DROP ROLE IF EXISTS sybase_fdw_superuser, sybase_fdw_user, no_priv_user;
RESET client_min_messages;
CREATE ROLE sybase_fdw_superuser LOGIN SUPERUSER; -- is a superuser
CREATE ROLE sybase_fdw_user LOGIN;                -- has priv and user mapping
CREATE ROLE no_priv_user LOGIN;                 -- has priv but no user mapping
-- Install sybase_fdw
CREATE EXTENSION sybase_fdw;
-- sybase_fdw_superuser owns fdw-related objects
SET ROLE sybase_fdw_superuser;
CREATE SERVER sybase_server FOREIGN DATA WRAPPER sybase_fdw;
-- privilege tests
SET ROLE sybase_fdw_user;
CREATE FOREIGN DATA WRAPPER sybase_fdw2 HANDLER sybase_fdw_handler VALIDATOR sybase_fdw_validator;   -- ERROR
ERROR:  permission denied to create foreign-data wrapper "sybase_fdw2"
HINT:  Must be superuser to create a foreign-data wrapper.
CREATE SERVER sybase_server2 FOREIGN DATA WRAPPER sybase_fdw;   -- ERROR
ERROR:  permission denied for foreign-data wrapper sybase_fdw
CREATE USER MAPPING FOR sybase_fdw_user SERVER sybase_server;   -- ERROR
ERROR:  permission denied for foreign server sybase_server
SET ROLE sybase_fdw_superuser;
GRANT USAGE ON FOREIGN SERVER sybase_server TO sybase_fdw_user;
SET ROLE sybase_fdw_user;
CREATE USER MAPPING FOR sybase_fdw_user SERVER sybase_server;
-- create user mappings and grant privilege to test users
SET ROLE sybase_fdw_superuser;
CREATE USER MAPPING FOR sybase_fdw_superuser SERVER sybase_server;
CREATE USER MAPPING FOR no_priv_user SERVER sybase_server;
